### PR TITLE
Ctor, dtor and finalizer body blocks

### PR DESCRIPTION
**Pull request by IIFE**
Add body blocks for ctor, dtor and finalizer to allow consumer to modify these methods bodies.

Create a ctor overload that takes a bool from the caller to tell the wrapper if it should own the native pointer passed to it, therefore making it responsible for deleting it later, or the caller is the only owner of the pointer.

Didn't think there were any tests needed as it is only generating a new ctor, which can be verified by looking at the generated code. How the ctor is used is up to the consumer.

### CODE REVIEW

**Comment by IIFE** [comment]
@ddobrev @tritao Thought I'd tag you in case this escaped your radar. Thanks.

File: src/Generator/Utils/BlockGenerator.cs [hunk 1]
```diff
@@ -50,6 +50,9 @@ public enum BlockKind
         Destructor,
         AccessSpecifier,
         Fields,
+        ConstructorBody,
```
Comment (tritao) [inline src/Generator/Utils/BlockGenerator.cs:53]: Do we need these explicit block kinds or could you just use the generic `MethodBody` one?

**Comment by tritao** [comment]
This looks mostly good, but I question if we really need these new block kinds.

Could you not just use `MethodBody` for tagging their body implementation parts?

Also something is wrong with the commits, you have some merge commits in there, would be great if you could clean those up from the PR.

File: src/Generator/Utils/BlockGenerator.cs (hunk 1 above)
Comment (IIFE) [reply src/Generator/Utils/BlockGenerator.cs:53]: `MethodBody` does not allow consumer to specifically know which method is being generated. These explicit blocks allow the consumer to detect when ctor, dtor or finalizer are generated.

**Comment by IIFE** [comment]
@tritao Want me to submit a new PR with cleaner commits?

File: src/Generator/Generators/CLI/CLIHeaders.cs [hunk 2]
```diff
@@ -376,6 +376,7 @@ public void GenerateClassConstructors(Class @class, string nativeType)
 
             // Output a default constructor that takes the native pointer.
             WriteLine("{0}({1} native);", @class.Name, nativeType);
+            WriteLine("{0}({1} native, const bool ownNativeInstance);", @class.Name, nativeType);
```
[the remark this answers is not in the document]
Comment (IIFE) [reply src/Generator/Generators/CLI/CLIHeaders.cs]: Done

File: src/Generator/Generators/CLI/CLIMarshal.cs [hunk 3]
```diff
@@ -295,15 +299,15 @@ public string QualifiedIdentifier(Declaration decl)
...
+        public void WriteClassInstance(Class @class, string instance, bool ownNativeInstance = false)
         {
             if (@class.IsRefType)
                 Context.Return.Write("({0} == nullptr) ? nullptr : gcnew ",
                     instance);
 
             Context.Return.Write("{0}(", QualifiedIdentifier(@class));
             Context.Return.Write("(::{0}*)", @class.QualifiedOriginalName);
-            Context.Return.Write("{0})", instance);
+            Context.Return.Write("{0}{1})", instance, !ownNativeInstance ? "" : ", true");
```
[the remark this answers is not in the document]
Comment (IIFE) [reply src/Generator/Generators/CLI/CLIMarshal.cs]: Done

**Comment by tritao** [comment]
You can force-push to this branch, no need for a new PR.

**Comment by IIFE** [comment]
I'm closing this in favour of #1318, because I can't seem to clean up the commits.

**Pull request closed without merging**